### PR TITLE
Jupyter notebook illustrating CONUS network analysis

### PR DESCRIPTION
**Pull request by awlostowski-noaa**
This notebook illustrates how t-route network analysis tools are used to decomposed a supernetwork into subnetworks and reaches. The notebook runs an example on the CONUS mainstem supernetwork. Maps are created along the way to visualize what is happening. 

## Additions

- A new notebook describing the use of network analysis functions

## Removals

- N/A

## Changes

- N/A

## Testing

1. Repeated execution of the notebook. 

## Screenshots
image

## Todos
- Ryan can revise this notebook to clarify some of the math/compsci lingo. I tried to stick to discussing concepts in terms of river network elements (segments, reaches, subnetworks, headwaters, tailwaters, etc.). It may be worthwhile to add more text that ties these terms to math-community terms (e.g. nodes, links, reverse network, etc.)
- Ryan can help make some of Adam's python scripting less jenky, particularly in the area of plotting.

### CODE REVIEW

File: notebooks/network-analysis.ipynb [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+    "# call function above...\n",
+    "data = replace_downstreams(data, network_data['downstream_col'], 0)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Extract network connections\n",
+    "- Upstream and downstream network connections are extracted with the `extract_network` and `reverse_network` functions\n",
```
Comment (groutr) [inline notebooks/network-analysis.ipynb:207]: This should read more clearly, "Downstream and upstream network connections ..."

File: notebooks/network-analysis.ipynb [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+    "- `extract_network` accepts the network DataFrame, `data`, and returns a dictionary of segments and their respective downstream connections. \n",
+    "- `reverse_network` flips the orientation of the network such the tailwaters are at the head of the network, and connections propagate into the headwaters. In doing so, each segments upstream connections are identified. The function accepts the `connections` dictionary, and returns a dictionary of segments and their respective upstream connections. "
```
Comment (groutr) [inline notebooks/network-analysis.ipynb:209]: reverse_network flips the edge orientation. The result is that headwaters in the original become tailwaters in the reversed network.

File: notebooks/network-analysis.ipynb [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Given the network DataFrame, `extract_network` returns a dictionary segments and downstream connections.\n",
+    "# segment-id: [downstream-connected-segment-id]\n",
+    "connections = nhd_network.extract_network(data, network_data[\"downstream_col\"])\n",
```
Comment (groutr) [inline notebooks/network-analysis.ipynb:220]: We should show that we are working with directed acyclic graph by showing there are no cycles in the graph. nhd_network has a topological sorting function that can check this (see the kahn_toposort function).  If that function can successfully sort the graph, we have a DAG and we can be certain that our routing assumptions will work.

File: notebooks/network-analysis.ipynb [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+   "source": [
+    "# Get the tailwater keys for each subnetwork \n",
+    "n_subnets = subreachable.keys()\n",
+    "\n",
+    "# initialize a new column in the GeoDataFrame that will containa subnetwork code\n",
+    "gdf[\"subnetwork\"] = 0\n",
+    "\n",
+    "# loop through subnetworks, populate the newly created \"subnetwork\" columm with an integer subnetwork identifier code\n",
+    "i = 1\n",
+    "for sub in n_subnets:\n",
```
Comment (groutr) [inline notebooks/network-analysis.ipynb:272]: This loop can be structured better (and probably faster too):
```python
for i, reached in enumerate(subreachable.values(), 1):
    gdf.loc[gdf.index.isin(reached), "subnetwork"] = i
```
The negative nodes shouldn't be appearing in the subreachable lists. They get filtered out by nhd_network.extract_network.

File: notebooks/network-analysis.ipynb [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Identify reaches\n",
+    "- Identifying reaches in each subnetwork is a two step process\n",
+    "    1. Use the `reachable_network` function to identify the upstream connectivity structure of each subnetwork. The function accepts the reverse network dictionary, `rconn`, and returns a dictionary, `subnets_`, which lists upstream connections for each segment in each subnetwork. \n",
+    "    2. Conduct a [depth-first-search](https://en.wikipedia.org/wiki/Depth-first_search) of each subnetwork. This is done with the `dfs_decomposition` function, which accepts the forward (downstream oriented) network dictionary, `nhd_network.reverse_network(net)`, and returns a dictionary of segment lists defining specific reaches in the subnetwork. Results from each subnetwork are concatenated into a larger dictionary, `subreaches`, which contains segment lists for all reaches in the supernetwork.    \n"
```
Comment (groutr) [inline notebooks/network-analysis.ipynb:312]: Big note here: dfs_decomposition doesn't reverse the network anymore as of yesterday. It operates on the network you pass in.

File: notebooks/network-analysis.ipynb [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# extract a single large subnetwork from the supernetwork GeoDataFrame.\n",
+    "# Turns out that subnetwork code 210 is the Colorado River Basin\n",
+    "C = gdf[\"subnetwork\"] == 210\n",
```
Comment (groutr) [inline notebooks/network-analysis.ipynb:359]: There should be a more robust way to do this. It's never guaranteed to always be code 210.

File: notebooks/network-analysis.ipynb [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+    "sub_gdf = gdf[C]\n",
+    "\n",
+    "# Turns out that the tailwater segment of the Colorado River Basin is 21412883\n",
+    "rchs = subreaches[-21412883]\n",
+    "\n",
+    "# create a reach-code column in the GeoDataFrame\n",
+    "sub_gdf.loc[:, \"reach\"] = 0\n",
```
Comment (groutr) [inline notebooks/network-analysis.ipynb:366]: This is simply `sub_gdf["reach"] = 0`

**Comment by awlostowski-noaa** [comment]
I am going to close this request because my notebook broke when I pulled some recent changes made in nhd_network.py. I ended up cloning an older version of the network2 branch prior to the 6/23 commits that changed nhd_network.py. I have created my own branch from this and will submit a pull request from there. I too am confused, but this is how I learn.

**Pull request closed without merging**